File: py110/small_problems_110_119/easy3/10.py
# https://launchschool.com/exercises/9e90e4a5


def is_balanced(string):
    stack = []

    for char in string:
        if not stack and char == ")":
            return False
        elif char == "(":
            stack.append(char)
        elif char == ")" and stack[-1] == "(":
            stack.pop()
    
    return not stack

print(is_balanced("What (is) this?") == True)        # True
print(is_balanced("What is) this?") == False)        # True
print(is_balanced("What (is this?") == False)        # True
print(is_balanced("((What) (is this))?") == True)    # True
print(is_balanced("((What)) (is this))?") == False)  # True
print(is_balanced("Hey!") == True)                   # True
print(is_balanced(")Hey!(") == False)                # True
print(is_balanced("What ((is))) up(") == False)      # True

File: js210/simpleProblems/easy3/10.js
// https://launchschool.com/exercises/5401a895

'use strict';

function wordSizes(string) {
  if (!string) return {};

  let replacedString = string.replace(/[^a-z\s]/gi, '');
  console.log(replacedString)
  return replacedString.split(' ')
                       .map(word => word.length)
                       .reduce((obj, wordLength) => {
                        obj[wordLength] ? obj[wordLength] += 1 : obj[wordLength] = 1;
                        return obj;
                       }, {});
}

console.log(wordSizes('Four score and seven.'));                       // { "3": 1, "4": 1, "5": 2 }
console.log(wordSizes('Hey diddle diddle, the cat and the fiddle!'));  // { "3": 5, "6": 3 }
console.log(wordSizes("What's up doc?"));                              // { "5": 1, "2": 1, "3": 1 }
console.log(wordSizes(''));                                            // {}